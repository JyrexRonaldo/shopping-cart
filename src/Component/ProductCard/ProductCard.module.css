.card {
  width: 300px;
  height: 450px;
  padding: 15px;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
}

.card > div:first-of-type {
  width: 300px;
  height: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.card img {
  width: 150px;
  height: auto;
  object-fit: contain;
}

.card > p:nth-of-type(1) {
  display: flexbox;
  align-items: center;
  text-align: center;
  height: 30px;
  width: max-width;
  overflow: hidden;
  text-overflow: ellipsis;
}

.card input {
  border: none;
  background-color: #f9f9f9;
  border-radius: 20px;
  width: 20%;
  padding: 0px 10px;
}

.card > div:nth-of-type(2) {
  display: flex;
  justify-content: center;
  gap: 5px;
}
