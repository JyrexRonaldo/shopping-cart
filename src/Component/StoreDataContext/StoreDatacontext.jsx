import { createContext } from "react";

const StoreDataContext = createContext(null);

export default StoreDataContext