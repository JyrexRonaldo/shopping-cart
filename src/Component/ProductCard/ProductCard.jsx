import { useContext, useState } from "react";
import styles from "./ProductCard.module.css";
import PropTypes from "prop-types";
import StoreDataContext from "../StoreDataContext/StoreDataContext";

function ProductCard({ imgURL, title, price, cardId }) {
  const [quantity, setQuantity] = useState(1);

  function handleQuantityInput(e) {
    setQuantity(e.target.value);
  }

  const { cartItems, setCartItems } = useContext(StoreDataContext);

  function handleAddToCart() {
    if (quantity < 1) {
      alert("Quantity can't be less than 1");
    } else if (!Object.keys(cartItems).includes(cardId.toString())) {
      setCartItems({ ...cartItems, ...{ [cardId]: +quantity } });
    } else {
      cartItems[cardId] += +quantity;
      setCartItems({ ...cartItems });
    }
    setQuantity(1);
  }

  return (
    <div className={styles.card}>
      <div>
        <img src={imgURL} alt="" />
      </div>
      <p>{title}</p>
      <p>{`$${price}`}</p>
      <div>
        Quantity:
        <input type="number" value={quantity} onInput={handleQuantityInput} />
      </div>
      <button type="button" onClick={handleAddToCart}>
        Add to Cart
      </button>
    </div>
  );
}

ProductCard.propTypes = {
  imgURL: PropTypes.string,
  title: PropTypes.string,
  price: PropTypes.number,
  cardId: PropTypes.number,
};

export default ProductCard;
