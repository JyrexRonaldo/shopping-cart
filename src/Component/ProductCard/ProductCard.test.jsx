import { render, screen } from "@testing-library/react";
import ProductCard from "./ProductCard";
import { describe, it, expect } from "vitest";
import StoreDataContext from "../StoreDataContext/StoreDataContext";
import userEvent from "@testing-library/user-event";

describe("ProductCard component", () => {
  it("it displays elements with props supplied", () => {
    const mockItemData = {
      imgURl: "image of item",
      title: "cool shirt",
      price: "429",
      card: 5,
    };

    const contextData = {};

    const { container } = render(
      <StoreDataContext.Provider value={contextData}>
        <ProductCard
          imgURL={mockItemData.imgURl}
          title={mockItemData.title}
          price={mockItemData.price}
          cardId={mockItemData.card}
        />
      </StoreDataContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("updates quantity input display after interaction", async () => {
    const user = userEvent.setup();

    const mockItemData = {
      imgURl: "image of item",
      title: "cool shirt",
      price: "429",
      card: 5,
    };

    const contextData = {};

    render(
      <StoreDataContext.Provider value={contextData}>
        <ProductCard
          imgURL={mockItemData.imgURl}
          title={mockItemData.title}
          price={mockItemData.price}
          cardId={mockItemData.card}
        />
      </StoreDataContext.Provider>
    );

    const quantityInput = screen.getByRole("spinbutton");

    await user.clear(quantityInput);
    await user.type(quantityInput, "5000");

    expect(quantityInput).toHaveValue(5000);
  });

  it("does not updates quantity input display after no interaction", () => {
    const mockItemData = {
      imgURl: "image of item",
      title: "cool shirt",
      price: "429",
      card: 5,
    };

    const contextData = {};

    render(
      <StoreDataContext.Provider value={contextData}>
        <ProductCard
          imgURL={mockItemData.imgURl}
          title={mockItemData.title}
          price={mockItemData.price}
          cardId={mockItemData.card}
        />
      </StoreDataContext.Provider>
    );

    const quantityInput = screen.getByRole("spinbutton");

    expect(quantityInput).toHaveValue(1);
  });
});
