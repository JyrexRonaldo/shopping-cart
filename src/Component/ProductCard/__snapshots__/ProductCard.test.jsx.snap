// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard component > input  1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > input field  1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > input field updates as expected 1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > input field updatses as  1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > input field updatses as expected 1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > it displays elements with props supplied 1`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;

exports[`ProductCard component > it displays elements with props supplied 2`] = `
<div>
  <div
    class="_card_371a27"
  >
    <div>
      <img
        alt=""
        src="image of item"
      />
    </div>
    <p>
      cool shirt
    </p>
    <p>
      $429
    </p>
    <div>
      Quantity:
      <input
        type="number"
        value="1"
      />
    </div>
    <button
      type="button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
